File: swift/Maze.swift
//
//  Maze.swift
//  maze-escape
//
//  Created by Wayne Howarth on 21/12/2022.
//

import Foundation

public class Maze {
	
	private var width: Int
	private var height: Int
	
	private var walls: [UInt8]
	
	public init(width: Int, height: Int) {
		self.width = width
		self.height = height
		
		walls = Array(repeating: Direction.None.rawValue, count: width * height)
		
		for i in 0..<width {
			makeWall(at: (i, 0), edge: .South)
			makeWall(at: (i, height - 1), edge: .North)
		}
		
		for j in 0..<height {
			makeWall(at: (0, j), edge: .West)
			makeWall(at: (width - 1, j), edge: .East)
		}
	}
	
	public func getWidth() -> Int {
		return self.width
	}
	
	public func getHeight() -> Int {
		return self.height
	}
	
	public func canMoveNorth(from: (Int, Int)) -> Bool {
		return !isWall(at: (from.0, from.1), direction: .North)
	}
	
	public func canMoveSouth(from: (Int, Int)) -> Bool {
		return !isWall(at: (from.0, from.1), direction: .South)
	}
	
	public func canMoveEast(from: (Int, Int)) -> Bool {
		return !isWall(at: (from.0, from.1), direction: .East)
	}
	
	public func canMoveWest(from: (Int, Int)) -> Bool {
		return !isWall(at: (from.0, from.1), direction: .West)
	}
	
	public func isWall(at: (Int, Int), direction: Direction) -> Bool {
		guard at.0 >= 0 && at.0 <= width - 1 && at.1 >= 0 && at.1 <= height - 1 else {
			return false
		}
		
		return walls[at.1 * width + at.0] & direction.rawValue != 0
	}
	
	public func makeWall(at: (Int, Int), edge: Direction) {
		guard at.0 >= 0 && at.0 <= width - 1 && at.1 >= 0 && at.1 <= height - 1 else {
			return
		}
		
		walls[at.1 * width + at.0] = walls[at.1 * width + at.0] | edge.rawValue
		
		if edge == .North && at.1 < height - 1 {
			walls[(at.1 + 1) * width + at.0] = walls[(at.1 + 1) * width + at.0] | edge.Opposite().rawValue
		}
		
		if edge == .South && at.1 > 0 {
			walls[(at.1 - 1) * width + at.0] = walls[(at.1 - 1) * width + at.0] | edge.Opposite().rawValue
		}
		
		if edge == .East && at.0 < width - 1 {
			walls[at.1 * width + at.0 + 1] = walls[at.1 * width + at.0 + 1] | edge.Opposite().rawValue
		}
		
		if edge == .West && at.0 > 0 {
			walls[at.1 * width + at.0 - 1] = walls[at.1 * width + at.0 - 1] | edge.Opposite().rawValue
		}
	}
}

File: scripts/Maze.ts
class Maze {
    private _width: number;
    private _height: number;
    private _walls: Direction[][];

    constructor(width: number, height: number) {
        this._width = width;
        this._height = height;
        this._walls = [];

        for (var i = 0; i < height; i++) {
            this._walls[i] = [];

            for (var j = 0; j < width; j++) {
                this._walls[i][j] = Direction.None;
            }
        }

        for (var i = 0; i < width; i++) {
            this.makeWall(i, 0, Direction.South);
            this.makeWall(i, height - 1, Direction.North);
        }

        for (var j = 0; j < height; j++) {
            this.makeWall(0, j, Direction.West);
            this.makeWall(width - 1, j, Direction.East);
        }
    }

    width(): number {
        return this._width;
    }

    height(): number {
        return this._height;
    }

    canMoveNorth(x: number, y: number): boolean {
        return !this.isWallAt(x, y, Direction.North);
    }

    canMoveSouth(x: number, y: number): boolean {
        return !this.isWallAt(x, y, Direction.South);
    }

    canMoveEast(x: number, y: number): boolean {
        return !this.isWallAt(x, y, Direction.East);
    }

    canMoveWest(x: number, y: number): boolean {
        return !this.isWallAt(x, y, Direction.West);
    }

    isWallAt(x: number, y: number, direction: Direction): boolean {
        if (x < 0 || x > this._width - 1 || y < 0 || y > this._height - 1)
            return false;

        return (this._walls[y][x] & direction) !== 0;
    }

    makeWall(x: number, y: number, edge: Direction) {
        if (x < 0 || x > this._width - 1 || y < 0 || y > this._height - 1 || edge === Direction.None)
            return;

        this._walls[y][x] = this._walls[y][x] | edge;

        if (edge === Direction.North && y < this._height - 1)
            this._walls[y + 1][x] = this._walls[y + 1][x] | this.opposite(edge);
        
        if (edge === Direction.South && y > 0)
            this._walls[y - 1][x] = this._walls[y - 1][x] | this.opposite(edge);

        if (edge === Direction.East && x < this._width - 1)
            this._walls[y][x + 1] = this._walls[y][x + 1] | this.opposite(edge);

        if (edge === Direction.West && x > 0)
            this._walls[y][x - 1] = this._walls[y][x - 1] | this.opposite(edge);
    }

    private opposite(edge: Direction): Direction {
        var _opposite: Record<Direction, Direction> = {
            [Direction.None]: Direction.None,
            [Direction.North]: Direction.South,
            [Direction.South]: Direction.North,
            [Direction.East]: Direction.West,
            [Direction.West]: Direction.East
        };

        return _opposite[edge];
    }
}